### PR TITLE
Add GPU support for batched hashing.

### PR DESCRIPTION
**Pull request by porcuquine**
- Add support for `ColumnTreeBuilder` which can be called to incrementally add columns then build a tree.
- `ColumnTreeBuilder` is configured with optional column and tree `Batcher`s.
-  If no `Batcher` is provided, ordinary CPU hashing is performed.
- `Batcher` has `CPU` and `GPU` variants
- `GPU` variant makes use of batched hashing (on GPU) of 2, 8, or 11-ary hashes.
-  When created,`Batcher` must be configured with max batch size.
- If max batch size is too large, hashing will exceed GPU RAM, so choose wisely (and test).
- With enough memory, GPU parallelism is the bottleneck, so there is no need to push max batch size past the point of best performance.
- `gpu` feature flag must be enabled to use `GPU`
- `gpu` currently unsupported on MacOS.
- Test GPU performance and experiment with batch sizes using `gbench` executable.
  - For now, batch sizes must be hardcoded (TODO: take CLI options).

Example run:
[NOTE: In the final version of this PR, it's necessary to change to the `gbench` directory before running, and the feature flag must not be provided.]
```
➜  neptune git:(gpu) RUST_LOG=info cargo run --release --features=gpu
   Compiling neptune v0.5.4 (/home/porcuquine/dev/neptune)
    Finished release [optimized] target(s) in 1.34s
     Running `target/release/gbench`
[2020-04-26T18:35:02Z INFO  gbench] KiB: 4194304
[2020-04-26T18:35:02Z INFO  gbench] leaves: 134217728
[2020-04-26T18:35:02Z INFO  gbench] max column batch size: 400000
[2020-04-26T18:35:02Z INFO  gbench] max tree batch size: 700000
[2020-04-26T18:35:02Z INFO  gbench] Creating ColumnTreeBuilder
[2020-04-26T18:35:06Z INFO  gbench] ColumnTreeBuilder created
[2020-04-26T18:35:06Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-04-26T18:35:06Z INFO  gbench] adding column batches
[2020-04-26T18:35:06Z INFO  gbench] start commitment
...............................................................................................................................................................................................................................................................................................................................................
[2020-04-26T18:37:18Z INFO  gbench] adding final column batch and building tree
[2020-04-26T18:37:34Z INFO  gbench] end commitment
[2020-04-26T18:37:34Z INFO  gbench] commitment time: 147.785702859s
[2020-04-26T18:37:34Z INFO  gbench] end
```

Bottom line: 11-ary column hashing and 8-ary tree building for 4GiB worth of sealing takes ~2m 28s.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,10 +13,16 @@ bellperson = "0.6.3"
 bincode = { version = "1.2", optional = true }
 blake2s_simd = "0.5"
 byteorder = "1"
+env_logger = "0.7.1"
```
Comment (dignifiedquire) [inline Cargo.toml:16]: You are adding quite a bit of dependencies here, that will never be actually used by the library. I would suggest either making them optional via a `feature` or moving your gbench program into its own subcrate

File: src/batch_hasher.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+use paired::bls12_381::Fr;
+use std::ops::Add;
+use typenum::bit::B1;
+use typenum::uint::{UInt, UTerm, Unsigned};
+
+#[derive(Clone, Debug)]
+pub enum BatcherType {
+    GPU,
+    CPU,
+}
```
Comment (dignifiedquire) [inline src/batch_hasher.rs:15]: why did you choose an `enum` over a `trait`?

File: src/column_tree_builder.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    <TreeArity as Add<B1>>::Output: ArrayLength<Fr>,
+{
+    pub leaf_count: usize,
+    data: Vec<Fr>,
+    /// Index of the first unfilled datum.
+    fill_index: usize,
+    column_constants: PoseidonConstants<Bls12, ColumnArity>,
+    tree_constants: PoseidonConstants<Bls12, TreeArity>,
+    pub column_batcher: Option<Batcher<'a, ColumnArity>>,
+    pub tree_batcher: Option<Batcher<'a, TreeArity>>,
```
Comment (dignifiedquire) [inline src/column_tree_builder.rs:41]: why are some of these fields public, and some not?

File: src/column_tree_builder.rs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        let start = self.fill_index;
+        let column_count = columns.len();
+        let end = start + column_count;
+
+        if end > self.leaf_count {
+            return Err(Error::Other("too many columns".to_string()));
+        }
+
+        match &mut self.column_batcher {
+            Some(batcher) => {
```
Comment (dignifiedquire) [inline src/column_tree_builder.rs:62]: it is a bit nicer to write 
```rust
match self.column_batcher {
  Some(ref mut batcher) => {},
  None => {}
}
```

File: src/column_tree_builder.rs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+    fn reset(&mut self) {
+        self.fill_index = 0;
+        self.data.iter_mut().for_each(|place| *place = Fr::zero());
+    }
+}
+fn as_generic_arrays<'a, Arity: ArrayLength<Fr>>(vec: &'a [Fr]) -> &'a [GenericArray<Fr, Arity>] {
+    assert_eq!(0, vec.len() % Arity::to_usize());
+
+    unsafe {
```
Comment (dignifiedquire) [inline src/column_tree_builder.rs:96]: please write a comment why this is safe

File: src/column_tree_builder.rs [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+        tree
+    }
+
+    fn reset(&mut self) {
+        self.fill_index = 0;
+        self.data.iter_mut().for_each(|place| *place = Fr::zero());
+    }
+}
+fn as_generic_arrays<'a, Arity: ArrayLength<Fr>>(vec: &'a [Fr]) -> &'a [GenericArray<Fr, Arity>] {
+    assert_eq!(0, vec.len() % Arity::to_usize());
```
Comment (dignifiedquire) [inline src/column_tree_builder.rs:94]: please add an assertion for the mem size of `Fr`

File: src/column_tree_builder.rs [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+                    while total_hashed < new_row_size {
+                        let batch_end = usize::min(batch_start + (max_batch_size * arity), row_end);
+                        let batch_size = (batch_end - batch_start) / arity;
+                        let preimages =
+                            as_generic_arrays::<TreeArity>(&tree_data[batch_start..batch_end]);
+                        let hashed = batcher.hash(&preimages)?;
+
+                        // Poor-man's copy_from_slice avoids a mutable second borrow of tree_data.
+                        for i in 0..hashed.len() {
+                            tree_data[new_row_start + total_hashed + i] = hashed[i]
```
Comment (dignifiedquire) [inline src/column_tree_builder.rs:183]: you should be able to make this work like this

```rust
let preimages = as_generic_arrays::<TreeArity>(&tree_data[batch_start..batch_end]);
let hashed = batcher.hash(&preimages)?;
drop(preimages); // make sure we don't reference tree_data anymore
tree_data[new_row_start + total_hashed..new_row_start + total_hashed + hashed.len()].copy_from_slice(hashed);
```

File: src/gpu.rs [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+use triton::{Array_u64_1d, Array_u64_2d, Array_u64_3d};
+use typenum::bit::B1;
+use typenum::{UInt, UTerm, Unsigned, U11, U2, U8};
+
+type P2State = triton::FutharkOpaqueP2State;
+type P8State = triton::FutharkOpaqueP8State;
+type P11State = triton::FutharkOpaqueP11State;
+pub(crate) type T864MState = triton::FutharkOpaqueT864MState;
+
+enum BatcherState {
```
Comment (dignifiedquire) [inline src/gpu.rs:19]: Can I get some more doc comments on the enums, structs and methods in this file?

File: src/gpu.rs [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+            Some(_) => Some(1 << 21), // Leaves for 64MiB tree. TODO: be more general.
+            None => None,
+        }
+    }
+
+    fn build_tree(&mut self, leaves: &[Fr]) -> Result<Vec<Fr>, Error> {
+        if let Some(state) = &self.tree_builder_state {
+            build_tree8_64m(&mut self.ctx, state, leaves)
+        } else {
+            panic!("Tried to build tree without tree_builder_state.");
```
Comment (dignifiedquire) [inline src/gpu.rs:116]: it is a bit nicer to write

```rust
let state = self.tree_builder_state.as_ref().expect("Tried to build without state");
// ... 
```

File: src/gpu.rs [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+    for (src, dest) in frs.iter().zip(res.chunks_mut(4)) {
+        dest.copy_from_slice(&src.into_repr().0);
+    }
+    res
+}
+
+fn frs_2d_to_u64s(frs_2d: &[Vec<Fr>]) -> Vec<u64> {
+    frs_2d
+        .iter()
+        .flat_map(|row| frs_to_u64s(row).into_iter())
```
Comment (dignifiedquire) [inline src/gpu.rs:183]: This is pretty inefficient, as it allocates the vectors and then throws them away again. Lets add at least a `TODO` to improve that

Comment (LuoZijun) [reply src/gpu.rs:183]: See https://github.com/filecoin-project/neptune/pull/14 .

File: src/batch_hasher.rs (hunk 2 above)
Comment (porcuquine) [reply src/batch_hasher.rs:15]: So users of `ColumnTreeBuilder` could pass a `BatcherType` and have the appropriate `Batcher` created for both column and tree building. I thought this would provide a simpler interface for callers who want to create multiple builders of various kinds depending on configuration information, in tests, etc. without having to support three different configurations (no batcher, GPU, CPU) as distinct types.

File: src/column_tree_builder.rs (hunk 7 above)
Comment (porcuquine) [reply src/column_tree_builder.rs:183]: Thank you. I am enriched.

File: src/gpu.rs (hunk 9 above)
Comment (porcuquine) [reply src/gpu.rs:116]: Thanks. I tried every other combination of characters several times.

File: src/gpu.rs (hunk 10 above)
Comment (porcuquine) [reply src/gpu.rs:183]: Awesome, thanks both.

File: src/batch_hasher.rs (hunk 2 above)
Comment (porcuquine) [reply src/batch_hasher.rs:15]: There is also trait (`BatchHasher`).

File: src/column_tree_builder.rs (hunk 3 above)
Comment (porcuquine) [reply src/column_tree_builder.rs:41]: I only needed `column_batcher` to be public, but user code might reasonably also want access to the `tree_batcher` to check the batch size, so I made it public too. The constants are only used internally and don't need to be exposed.

File: Cargo.toml [hunk 11]
```diff
@@ -10,19 +10,19 @@ repository = "https://github.com/porcuquine/poseidon"
...
 bellperson = "0.6.3"
-bincode = { version = "1.2", optional = true }
 blake2s_simd = "0.5"
 byteorder = "1"
 ff = { version = "0.2.1", package = "fff" }
 generic-array = "0.13.2"
-paired = "0.17.0"
-rand_xorshift = "0.2.0"
+paired = "0.18.0"
+triton = { version = "0.2.0", package = "neptune-triton", default-features=false, features=["opencl"] }
```
Comment (dignifiedquire) [inline Cargo.toml:18]: this should be an optional dependency, which is activated with the gpu feature
